destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala

```diff
@@ -0,0 +1,221 @@
+      val uName = hyName.unsafeForSqlName
+      val uSchema = hySchema.unsafeForSqlName
+
+      new TempTable {
+        def ingest(chunk: Chunk[CharSequence]): ConnectionIO[Unit] =
+          insertChunk(log)(tempFragment, columns, chunk)
+
+        def drop: ConnectionIO[Unit] =
+          ifExists(log)(uName, uSchema.some).option flatMap { results =>
+            if (results.exists(_ === 1)) {
+              (fr"DROP TABLE" ++ tempFragment)
+                .updateWithLogHandler(log)
+                .run
+                .void
+            } else {
+              ().pure[ConnectionIO]
+            }
+          }
+
+        def create: ConnectionIO[Unit] =
+          ifExists(log)(uName, uSchema.some).option flatMap { results =>
+            if (results.exists(_ === 0)) {
+              val columnsObj = createColumnSpecs(columns)
+              (fr"CREATE TABLE" ++ tempFragment ++ fr0" " ++ columnsObj)
+                .updateWithLogHandler(log)
+                .run
+                .void
+            } else {
+              ().pure[ConnectionIO]
+            }
+          }
+
+        private def truncate: ConnectionIO[Unit] =
+          ifExists(log)(uName, uSchema.some).option flatMap { results =>
+            if (results.exists(_ === 1)) {
+              (fr"TRUNCATE TABLE" ++ tempFragment)
+                .updateWithLogHandler(log)
+                .run
+                .void
+            } else {
+              ().pure[ConnectionIO]
+            }
+          }
+
+        private def insertInto: ConnectionIO[Unit] =
+          (fr"INSERT INTO" ++
+            tableFragment ++ fr0" " ++
+            fr"SELECT * FROM" ++
+            tempFragment)
+            .updateWithLogHandler(log)
+            .run
+            .void
+
+        private def rename: ConnectionIO[Unit] =
+          (fr0"EXEC SP_RENAME '" ++
+            tempFragment ++ fr0"', '" ++
+            Fragment.const0(SQLServerHygiene.hygienicIdent(Ident(unsafeName)).unsafeForSqlName) ++ fr0"'")
+            .updateWithLogHandler(log)
+            .run
+            .void
+
+        private def filterTempIds(idColumn: Column[_]): ConnectionIO[Unit] = {
+          val mkColumn: String => Fragment = parent =>
+            Fragment.const0(parent) ++ fr0"." ++
+            Fragment.const0(SQLServerHygiene.hygienicIdent(Ident(idColumn.name)).forSqlName)
+
+          val fragment =
+            fr"DELETE target FROM" ++ tableFragment ++
+            fr" target INNER JOIN" ++ tempFragment ++ fr" temp" ++
+            fr"ON" ++ mkColumn("target") ++ fr0"=" ++ mkColumn("temp")
+
+          fragment.updateWithLogHandler(log).run.void
+        }
+
+        def append: ConnectionIO[Unit] = {
+          val mbFilter = idColumn traverse_ { col => filterTempIds(col) }
+          mbFilter >>
+          insertInto >>
+          truncate
+        }
+
+        def persist(writeMode: WriteMode): ConnectionIO[Unit] = {
```

Comment:
We might consider moving the `WriteMode` argument to the constructor, with that information we could choose to use [temporary tables](https://docs.microsoft.com/en-us/sql/t-sql/statements/create-table-transact-sql?view=sql-server-ver15#temporary-tables) when possible. We'd also be able to fail faster in the `Create` case.

Reply:
The problem with temp tables is that we _ptobably_ want to resume push even if sql server is restarted which will create different session (if I understand session correctly).

I don't mind moving write mode to constructor, but not sure how would it help. Could you elaborate please?

Reply:
> we probably want to resume push even if sql server is restarted

Are you referring to a restart during a push or in between pushes? It seems like the only time _during_ would really be significant is during the initial load of an incremental push, but there we'd be performing a replace/create so we wouldn't use a temp table anyhow. For subsequent loads of an incremental, we'd be resuming from previous so is it worth it to try and resume mid push, or just restart it?

As for what I meant about "failing faster", if we know the write mode ahead of time, we could check for the existence of the destination table in the "create" case before consuming anything (whereas now we may consume everything and then fail during `persist` b/c the table already exists).

Reply:
E.g. during a push connection was reset, we catch the error and retry insert chunk after timeout. Or we don't simply rollback on error during rename but again retry. 

Right, I'll add a check that table exists or not exists. Not sure about this though, as it's not locked and user might delete or create it during loading.

Reply:
> during a push connection was reset, we catch the error and retry insert chunk after timeout. Or we don't simply rollback on error during rename but again retry.

Yeah, is that what happens now (we retry)? But you're correct (afaict) that any temp tables would be gone, so we'd have to weigh the benefits.

> it's not locked and user might delete or create it during loading

Yeah, it is mostly to fail-fast in the common case. If it doesn't exist at the start and the user adds a conflicting table before we attempt to create, the create will still fail.